### PR TITLE
#36 Support selectable digest when create timestamp.

### PR DESCRIPTION
**Pull request by JunOkumura**
I implemented my just idea for #36 it is issue myself.

How about is this?
Could you tell me your idea?

### CODE REVIEW

File: lib/glueby/contract/timestamp.rb [hunk 1]
```diff
@@ -55,12 +55,14 @@ def initialize(
         wallet:,
         content:,
         prefix: '',
-        fee_provider: Glueby::Contract::FixedFeeProvider.new
+        fee_provider: Glueby::Contract::FixedFeeProvider.new,
+        digest: 'sha256'
```
Comment (azuchi) [inline lib/glueby/contract/timestamp.rb:59]: The digest value is better for symbols than strings like (:sha256, :double_sha256 or :none)

File: lib/glueby/contract/timestamp.rb [hunk 2]
```diff
@@ -70,9 +72,23 @@ def initialize(
...
 
-        @tx = create_tx(@wallet, @prefix, Tapyrus.sha256(@content), @fee_provider)
+        content = digest_content(@digest)
+        @tx = create_tx(@wallet, @prefix, content, @fee_provider)
         @txid = @wallet.internal_wallet.broadcast(@tx)
       end
+
+      private
+
+      def digest_content(digest = '')
```
Comment (azuchi) [inline lib/glueby/contract/timestamp.rb:82]: Since it is initialized in the constructor, you probably don't need the default values for the arguments.

**Review by azuchi — CHANGES_REQUESTED** [review]

File: lib/glueby/contract/timestamp.rb (hunk 1 above)
Comment (JunOkumura) [reply lib/glueby/contract/timestamp.rb:59]: OK

File: lib/glueby/contract/timestamp.rb (hunk 2 above)
Comment (JunOkumura) [reply lib/glueby/contract/timestamp.rb:82]: That's right. I will fix it.

I have a question.
What do it if digest is gave not supported symbol?

Comment (azuchi) [reply lib/glueby/contract/timestamp.rb:82]: It should raise error. Errors should be defined in `lib/glueby/contract/errors.rb`.

Comment (JunOkumura) [reply lib/glueby/contract/timestamp.rb:82]: OK, Thank you.

File: spec/glueby/contract/timestamp_spec.rb [hunk 3]
```diff
@@ -83,5 +83,60 @@
...
+        )
+      end
+
+      it 'create transaction for content that is not digested' do
+        subject
+        expect(contract.tx.inputs.size).to eq 1
+        expect(contract.tx.outputs.size).to eq 2
+        expect(contract.tx.outputs[0].value).to eq 0
+        expect(contract.tx.outputs[0].script_pubkey.op_return?).to be_truthy
+        expect(contract.tx.outputs[0].script_pubkey.op_return_data.bth).to eq '10'
```
Comment (azuchi) [inline spec/glueby/contract/timestamp_spec.rb:103]: data should be `01`.

**Review by azuchi — CHANGES_REQUESTED** [review]
In addition, the name of the `data_hash` argument of `Timestamp::Util#create_tx` is incorrect because it allows none in the digest.

File: spec/glueby/contract/timestamp_spec.rb [hunk 4]
```diff
@@ -83,5 +83,60 @@
 
       it { expect { subject }.to raise_error(Glueby::Contract::Errors::TxAlreadyBroadcasted) }
     end
+
+    context 'if digest is :none' do
+      let(:contract) do
+        Glueby::Contract::Timestamp.new(
+          wallet: wallet,
+          content: "\01",
```
Comment (azuchi) [inline spec/glueby/contract/timestamp_spec.rb:91]:
```suggestion
          content: "01",
```

File: spec/glueby/contract/timestamp_spec.rb [hunk 5]
```diff
@@ -83,5 +83,60 @@
...
+        expect(contract.tx.outputs[0].script_pubkey.op_return_data.bth).to eq '10'
+        expect(contract.tx.outputs[1].value).to eq 99_990_000
+      end
+    end
+
+    context 'if digest is :double_sha256' do
+      let(:contract) do
+        Glueby::Contract::Timestamp.new(
+          wallet: wallet,
+          content: "\01",
```
Comment (azuchi) [inline spec/glueby/contract/timestamp_spec.rb:112]:
```suggestion
          content: "01",
```

File: spec/glueby/contract/timestamp_spec.rb [hunk 6]
```diff
@@ -83,5 +83,60 @@
...
+        expect(contract.tx.outputs[0].script_pubkey.op_return_data.bth).to eq '9c12cfdc04c74584d787ac3d23772132c18524bc7ab28dec4219b8fc5b425f70'
+        expect(contract.tx.outputs[1].value).to eq 99_990_000
+      end
+    end
+
+    context 'if digest unspport' do
+      let(:contract) do
+        Glueby::Contract::Timestamp.new(
+          wallet: wallet,
+          content: "\01",
```
Comment (azuchi) [inline spec/glueby/contract/timestamp_spec.rb:133]:
```suggestion
          content: "01",
```

**Review by azuchi — CHANGES_REQUESTED** [review]

**Comment by JunOkumura** [comment]
OK, I will update them soon.

**Comment by azuchi** [comment]
@JunOkumura thanks. The rest is one.

> In addition, the name of the data_hash argument of Timestamp::Util#create_tx is incorrect because it allows none in the digest.

**Comment by JunOkumura** [comment]
Oh, sorry. I forgot it.
I will fix it soon.

**Review by azuchi — APPROVED** [review]

**Pull request merged**